### PR TITLE
nixos: Fix "Argument list too long" error

### PR DESCRIPTION
**Pull request by linyinfeng**
I encountered a weird "Argument list too long" error when building system configuration.

```
# nixos-rebuild --flake /etc/nixos build -vvvv --show-trace
...
... long regular outputs ...
...
building '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv'...
building of '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv!out' from .drv file: got EOF
building of '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv!out' from .drv file: woken up
building of '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv!out' from .drv file: build done
killing process 2625
builder process for '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv' finished
killing all processes running under uid '30001'
lock released on '/nix/store/6yf0nssx29rssphzr1ns94pglxgy2ik3-nixos-system-yinfeng-work-21.03.20201109.b839d4a.lock'
building of '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv!out' from .drv file: done
building of '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv!out' from .drv file: goal destroyed
error: --- Error -------------------------------------------------------------------------------------------------------------------------------------------------------------- nix
builder for '/nix/store/pgnfd73mkzyzzz613lahx94hgpsri557-nixos-system-yinfeng-work-21.03.20201109.b839d4a.drv' failed with exit code 1; last 2 log lines:
  
  ErroSysError{executing '/nix/store/qdp56fi357fgxxnkjrwx1g67hrk775im-bash-4.4-p23/bin/bash': Argument list too long
```

Because the configuration have over 50 directories in `environment.persistence."/persist".directories`...

Seems the issue was caused by a too long activation script. It can be easily fixed by writing the generated script into nix store. This fix is borrowed from [NixOS/nixops/pull/915](https://github.com/NixOS/nixops/pull/915).

### CODE REVIEW

**Review by cole-h — APPROVED** [review]
Diff LGTM.

**Review by talyz — CHANGES_REQUESTED** [review]
Thanks for working on a solution to this!

I would prefer if the script part was split out into its own dep string and the arguments to it were passed as regular bash arguments. That would simplify things and not put ~50 nearly identical files into the nix store.

**Comment by linyinfeng** [comment]
@talyz Thank you for your suggestion. I have added two commits for this PR. Now the script part will be written into store and the activation script should looks like this (test on my machine):

```bash
$ cat /run/current-system/activate
...
#### Activation script snippet createDirsIn--persist:
_localstatus=0
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/etc/nixos"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/etc/NetworkManager/system-connections"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/log"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/lib/docker"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/lib/flatpak"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/lib/bluetooth"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/lib/libvirt"
/nix/store/6a28cn56k9i8cfazs9bd4f45gchgv4fq-impermanence-create-directories "/persist" "/var/lib/systemd/coredump"
...
```

BTW, actually **the old fix** only create 1 file per persistence storage location per generation, but the file will have over 50 identical sections of codes.

Also, this TODO,

```bash
# TODO: Move this script to it's own file, add CI with shfmt/shellcheck.
```

can be performed easily if this PR is merged.

Finally, CI of this repository is broken now...

```
> Run cachix/install-nix-action@v9
...
Error: Unable to process command '::add-path::/nix/var/nix/profiles/per-user/runner/profile/bin' successfully.
Error: The `add-path` command is disabled. Please upgrade to using Environment Files or opt into unsecure command execution by setting the `ACTIONS_ALLOW_UNSECURE_COMMANDS` environment variable to `true`. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
...
```

**Comment by lovesegfault** [comment]
Please rebase to pick up the CI fixes :)

File: nixos.nix [hunk 1]
```diff
@@ -87,11 +87,11 @@ in
 
     system.activationScripts =
       let
-        # Create a directory in persistent storage, so we can bind
+        # Script to create a directory in persistent storage, so we can bind
         # mount it. The directory structure's mode and ownership mirror those of
         # persistentStoragePath/dir;
         # TODO: Move this script to it's own file, add CI with shfmt/shellcheck.
-        mkDirWithPerms = persistentStoragePath: dir:
+        create-directories = pkgs.writeShellScript "impermanence-create-directories"
```
Comment (talyz) [inline nixos.nix:94]:
```suggestion
        createDirectories = pkgs.writeShellScript "impermanence-create-directories"
```
To keep it in line with the `nixpkgs` variable naming style.

**Review by talyz — CHANGES_REQUESTED** [review]
Looks good! I only have a small change request. When that's addressed it would be great if you could rebase and squash all commits into one; then I think we can merge :+1: 

> BTW, actually the old fix only create 1 file per persistence storage location per generation ...

Ah, sorry, I assumed it would  create one file per directory to cope with the script length, but I should have read it more thoroughly :sweat_smile:

**Review by lovesegfault — APPROVED** [review]

**Review by talyz — APPROVED** [review]

**Comment by talyz** [comment]
Thanks!

**Pull request merged**